### ISSUE TITLE
tests leave behind unremovable file.

### ISSUE THREAD

**Issue by gbaz** [haskell/directory#194]
After resolving #97 (i think), running the tests for directory leaves a file that cannot be removed except by root.

This caused the hackage builder to stall for roughly a week because it was unable to remove the following file:

   `/home/builder/builder-dir/build-cache/tmp-install/directory-1.3.9.0/dist/test-RemovePathForcibly.tmp/removePathForcibly.tmp/f`

It would be good if the test suite were fixed to avoid this issue.

**Comment by Rufflewind** [comment]
Were the tests interrupted or did any of them fail? Under normal circumstances, RemovePathForcibly.hs is supposed to remove it successfully and there are assertions to verify the removal.

**Comment by gbaz** [comment]
they failed but the available logs don't explain why -- https://hackage.haskell.org/package/directory-1.3.9.0/reports/1

i do think its bad practice for a failing test suite to leave a file that needs to be fixed by hand.

**Comment by gbaz** [comment]
Reran on the builder and here are some lines in the log that seem suspicious:

> 
> *** LongPaths: exception: yeah_its_longyeah_its_longyeah_its_longyeah_its_longyeah_its_longyeah_its_longyeah_its_longyeah_its_long: removePathForcibly:openFdAt: permission denied (Permission denied)
> 
> *** RemovePathForcibly: exception: f: removePathForcibly:openFdAt: permission denied (Permission denied)
> 

The failures appear to be precisely that removePathForcibly does not work!

**Comment by gbaz** [comment]
Indeed, I can't remove the created file by hand as the `builder` user even with `rm -f` -- I need to change to superuser to do it.

**Comment by Rufflewind** [comment]
Turns out, the flag `AT_SYMLINK_NOFOLLOW` in `fchmodat` has [inconsistent support](https://lwn.net/Articles/939217/) among different permutations of kernels and C libraries. With this code,

https://github.com/haskell/directory/blob/dfe5e23f15076e8be01797e3252ee992a7ba4d97/main.c#L1-L19

testing indicates that:

- It is [supported](https://github.com/Rufflewind/directory/actions/runs/20685998807/job/59386714237) on the [macos-15](https://github.com/actions/runner-images/blob/main/images/macos/macos-15-arm64-Readme.md) and [ubuntu-24.04](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md) images of GitHub Actions workers.
- It is unsupported on my WSL 2.6.3.0 (kernel 6.6.87.2-1) and fails with `errno = 95` (`EOPNOTSUPP`).

This failure explains why `removePathForcibly` was unable to remove write-protected files on some systems, but cannot be reproduced on GitHub Actions workers.
